feat(HTTP): Refresh 5XX response pages (#35361)
mdn/content: files/en-us/web/http/status/506/index.md

{{HTTPSidebar}}

The HTTP **`506 Variant Also Negotiates`** [server error response](/en-US/docs/Web/HTTP/Status#server_error_responses) status code is returned during Transparent Content Negotiation (TCN) when there is recursive loop in the process of selecting a resource.

Transparent Content Negotiation enables a client and server to collaboratively decide the best variant of a given resource when the server has multiple variants.

> A server sends a `506` status code due to server misconfiguration when there are circular references when creating choice responses.

Reviewer: ```suggestion
A server sends a `506` status code due to server misconfiguration that results in circular references when creating responses.
```

Transparent Content Negotiation is not supported in most modern browsers due to complexity in implementations, lack of standardization of how clients automatically choose from responses, and the additional round-trips that slow down client-server interaction.

> [Server-driven content negotiation](/en-US/docs/Web/HTTP/Content_negotiation#server-driven_content_negotiation) is instead common, where a server directly chooses the most appropriate resource for the client based on the request headers ({{HTTPHeader("Accept-Language")}}, {{HTTPHeader("Accept")}}, etc.).

Reviewer: ```suggestion
[Server-driven content negotiation](/en-US/docs/Web/HTTP/Content_negotiation#server-driven_content_negotiation) is far more common, where a server directly chooses the most appropriate resource for the client based on the request headers ({{HTTPHeader("Accept-Language")}}, {{HTTPHeader("Accept")}}, etc.).
```

{{HTTPSidebar}}

The HTTP **`506 Variant Also Negotiates`** [server error response](/en-US/docs/Web/HTTP/Status#server_error_responses) status code may be given in the context of [content negotiation](/en-US/docs/Web/HTTP/Content_negotiation).
The pattern of content negotiation enables a client to retrieve the best variant of a given resource, where the server supports multiple variants.

A `506` status code indicates an internal server configuration error in which the chosen variant is itself configured to engage in content negotiation.

> To avoid this misconfiguration, final resources or files should be target resources and not participate in content negotiation to avoid recursive conflicts.

Reviewer: All of this is complete gibberish. I don't understand what the status code tells the browser, and what the browser should do next, or instead.

Reviewer: Note, it is no worse than what was there before though, so I would allow to merge on this, but we should fix it as a post process if we don't fix it now.

Author: Let's have a crack at it:

>you might get this error back during content negotiation where the server supports multiple variants of a resource but [for some reason that isn't clearly explained]

In Apache you do content negotiation with 

* a type map **or**
* Multiviews Options

https://httpd.apache.org/docs/trunk/mod/mod_negotiation.html

Where the type map (document.html.var) looks like:

```
URI: document.html

Content-language: en
Content-type: text/html
URI: document.html.en

Content-language: fr
Content-type: text/html
URI: document.html.fr
```

And if you request `document.html.var` with `Accept-language: fr`, you get `document.html.fr`.

If you do something wonky like:

```
URI: document.html

Content-language: fr
Content-type: text/html
URI: document.html.var
```

Then you will get a circular reference and Apache returns a 506.

Author: More context when trying to reproduce: this is specifically Transparent Content Negotiation (TCN) kicked off when the client request has `Negotiate: trans` header. If the variant chosen has `TCN: *` then it is itself engaging in TCN. None of this is supported in browsers anymore in favor of server-driven conneg where the server decides the best resource instead of going through another back-and-forth.

Author: https://httpd.apache.org/docs/trunk/content-negotiation.html#about

>httpd supports 'server driven' content negotiation, as defined in the HTTP/1.1 specification. It fully supports the Accept, Accept-Language, Accept-Charset and Accept-Encoding request headers. httpd also supports 'transparent' content negotiation, which is an experimental negotiation protocol defined in RFC 2295 and RFC 2296. It does not offer support for 'feature negotiation' as defined in these RFCs.

Reviewer: Thanks for the digging. What you have done is much better.
